src/iop/filmicrgb.c

```diff
@@ -1502,6 +1504,363 @@ static inline void filmic_chroma_v2_v3(const float *const restrict in, float *co
+                                 const dt_colormatrix_t input_matrix, const dt_colormatrix_t output_matrix,
+                                 const dt_colormatrix_t export_input_matrix, const dt_colormatrix_t export_output_matrix,
+                                 const float display_black, const float display_white, const float saturation,
+                                 const int use_output_profile)
+{
+  // Force final hue to original
+  Ych_final[2] = Ych_original[2];
+
+  // Clip luminance
+  Ych_final[0] = CLAMP(Ych_final[0],
+                        CIE_Y_1931_to_CIE_Y_2006(display_black),
+                        CIE_Y_1931_to_CIE_Y_2006(display_white));
+
+  // Massage chroma
+  filmic_desaturate_v4(Ych_original, Ych_final, saturation);
+  gamut_check_Yrg(Ych_final);
+
+  if(!use_output_profile)
+  {
+    // Gamut-map against pipeline RGB
+    Ych_to_pipe_RGB(Ych_final, output_matrix, pix_out);
+
+    // Now, it is still possible that one channel > display white because of saturation.
+    // We have already clipped Y, so we know that any problem now is caused by c
+    gamut_check_RGB(pix_out, Ych_final, input_matrix, output_matrix, display_black, display_white);
+  }
+  else
+  {
+    // Gamut-map against output RGB
+    Ych_to_pipe_RGB(Ych_final, export_output_matrix, pix_out);
+
+    // Now, it is still possible that one channel > display white because of saturation.
+    // We have already clipped Y, so we know that any problem now is caused by c
+    gamut_check_RGB(pix_out, Ych_final, export_input_matrix, export_output_matrix, display_black, display_white);
+
+    // Go from export RGB to CIE LMS 2006 D65
+    dt_aligned_pixel_t LMS = { 0.f };
+    dot_product(pix_out, export_input_matrix, LMS);
+
+    // Go from CIE LMS 2006 D65 to pipeline RGB D50
+    dot_product(LMS, output_matrix, pix_out);
+  }
+}
+
+
+static int filmic_v4_prepare_matrices(dt_colormatrix_t input_matrix, dt_colormatrix_t output_matrix,
+                                       dt_colormatrix_t export_input_matrix, dt_colormatrix_t export_output_matrix,
+                                       const dt_iop_order_iccprofile_info_t *const work_profile,
+                                       const dt_iop_order_iccprofile_info_t *const export_profile)
+
+{
+  dt_colormatrix_t temp_matrix;
+
+  // Prepare the pipeline RGB (D50) -> XYZ D50 -> XYZ D65 -> LMS 2006 matrix
+  dt_colormatrix_mul(temp_matrix, XYZ_D50_to_D65_CAT16, work_profile->matrix_in);
+  dt_colormatrix_mul(input_matrix, XYZ_D65_to_LMS_2006_D65, temp_matrix);
+
+  // Prepare the LMS 2006 -> XYZ D65 -> XYZ D50 -> pipeline RGB matrix (D50)
+  dt_colormatrix_mul(temp_matrix, XYZ_D65_to_D50_CAT16, LMS_2006_D65_to_XYZ_D65);
+  dt_colormatrix_mul(output_matrix, work_profile->matrix_out, temp_matrix);
+
+  // If the pipeline output profile is supported (matrix profile), we gamut map against it
+  const int use_output_profile = (export_profile != NULL);
```

Comment:
For https://github.com/darktable-org/darktable/pull/10975 I'm considering falling back to sRGB / Rec.709 if the export profile isn't a matrix profile. Do you think it would be a sane default here as well? I think the behaviour should be unified. Perhaps a control log message should also be emitted if an unsupported export profile is used.

Reply:
The sane default would be to default to working RGB, because in the case of printing profiles, the gamut is typically larger than sRGB/Rec709 but they are typically defined as LUT ICC profiles, which we don't handle internally. So, our best bet is that Little CMS2 will do the right thing later on.

Reply:
Yes, fair point. I agree.